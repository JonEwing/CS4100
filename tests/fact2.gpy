(fun fact (x i32) -> i32
  (let a (alloc 100 0) 
       (seq (set a 0 x)
            (let y (get a 0)
                   (cond (== 0 y) 1 (* y (fact (- y 1))))))))

%

(fact 12)