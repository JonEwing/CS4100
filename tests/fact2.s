setframe 0
push Lmain
call
halt
Lmain:
push 12
push Lfact
setframe 2
swap
call
ret
Lfact:
push undef
push undef
push 100
push 0
alloc
store 4
var 4
push 0
var 0
set
var 4
push 0
get
store 3
var 3
push 0
binary ==
push _L1
branch
push 1
var 3
binary -
push Lfact
setframe 2
swap
call
var 3
binary *
push true
push _L2
branch
_L1:
push 1
_L2:
store 3
pop
ret
