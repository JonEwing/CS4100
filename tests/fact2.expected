Vi32(479001600)